### PR TITLE
Features made for Bartercas3

### PR DESCRIPTION
**Pull request by boudewijn-tribler**
See commit descriptions.

### CODE REVIEW

File: authentication.py [hunk 1]
```diff
@@ -307,12 +307,19 @@ def set_signature(self, member, signature):
 
             @param signature: The signature for this message.
             @type signature: string
+
+            @param sign: Whether signatures should be created when we have the private key
+             available.
+            @type sign: bool
```
Comment (NielsZeilemaker) [inline authentication.py:314]: Maybe add documentation for the regenerate param

File: dispersy.py [hunk 2]
```diff
@@ -94,15 +94,21 @@ def create_identifier(number):
         assert isinstance(number, (int, long)), type(number)
         return u"request-cache:signature-request:%d" % (number,)
 
-    def __init__(self, request_cache, members, response_func, response_args, timeout):
+    def __init__(self, request_cache, candidates, members, response_func, response_args, success_func, success_args, timeout):
         super(SignatureRequestCache, self).__init__(request_cache)
         self.request = None
+        # CANDIDATES is a list containing all the candidates that should add their signature.
```
Comment (NielsZeilemaker) [inline dispersy.py:118]: Maybe add an assert?

File: dispersy.py [hunk 3]
```diff
@@ -2553,48 +2560,59 @@ def on_introduction_request(self, messages):
...
-            if direction == u"ASC":
-                return u"""
- SELECT * FROM
-  (SELECT sync.packet FROM sync    -- """ + meta.name + """
-   WHERE sync.meta_message = ? AND sync.undone = 0 AND sync.global_time BETWEEN ? AND ? AND (sync.global_time + ?) % ? = 0
-   ORDER BY sync.global_time ASC)"""
+            assert isinstance(meta.distribution, SyncDistribution), type(meta)
+            order_by = {u"ASC":"sync.global_time ASC", u"DESC":"sync.global_time DESC", u"RANDOM":"RANDOM()"}
+
+            if meta.distribution.members:
```
Comment (NielsZeilemaker) [inline dispersy.py:2602]: The pivot and modulo heuristic need to be modified to check the members as well, as the bloomfilters it is creating are checked by this method. Or the include_inactive param needs to be renamed and used to not perform a sub_select whenever we create an outgoing bloomfilter.

File: authentication.py (hunk 1 above)
Comment (boudewijn-tribler) [inline authentication.py:314]: Done :)

**Comment by boudewijn-tribler** [comment]
As discussed with @whirm, the branch https://github.com/boudewijn-tribler/dispersy/tree/merge-from-dimitra-BF-creation contains the work in progress to add destination specific bloom filter creation.  He will continue on this feature.

**Pull request closed without merging**